File: quant-perps-core/src/venues/paper.ts
import {
  type AccountState,
  type Order,
  type OrderResult,
  type Position,
  type Quote,
  type Side,
  opposite,
  posNotional,
  posUnrealizedPnl,
  sideSign,
} from "../types.js";
import type { PriceSource } from "./base.js";

// Deterministic perp simulation backing any venue in paper mode. Nets
// positions, averages entry on adds, realises PnL on reductions/flips, applies
// taker slippage, and reports a margin-style account. Mirrors the Python broker.
export class PaperBroker {
  cash: number;
  realizedPnl = 0;
  private positions = new Map<string, Position>();

  constructor(
    public readonly venue: string,
    private readonly priceSource: PriceSource,
    startingEquity = 100_000,
    private readonly slippageBps = 2,
  ) {
    this.cash = startingEquity;
  }

  getQuote(symbol: string): Quote | null {
    return this.priceSource.quote(symbol);
  }

  private fillPrice(order: Order, quote: Quote): number {
    // Marketable orders take liquidity at the book (ask for buys, bid for
    // sells); a limit is then capped so the fill is never worse than authorised.
    const base = order.side === "long" ? quote.ask : quote.bid;
    const slip = base * (this.slippageBps / 1e4);
    let fill = order.side === "long" ? base + slip : base - slip;
    if (order.orderType === "limit" && order.limitPrice !== undefined) {
      fill =
        order.side === "long"
          ? Math.min(fill, order.limitPrice)
          : Math.max(fill, order.limitPrice);
    }
    return fill;
  }

  placeOrder(order: Order): OrderResult {
    const quote = this.getQuote(order.symbol);
    if (!quote)
      return this.rejected(order, `no price for ${order.symbol}`);

    const existing = this.positions.get(order.symbol);
    if (order.reduceOnly && !existing)
      return this.rejected(order, "reduce_only order with no open position");

    // Non-marketable limits rest (PENDING) instead of filling at an impossible
    // price. We don't model a resting book, so they simply apply no fill.
    if (order.orderType === "limit" && order.limitPrice !== undefined) {
      const marketable =
        order.side === "long"
          ? order.limitPrice >= quote.ask
          : order.limitPrice <= quote.bid;
      if (!marketable)
        return {
          order,
          status: "pending",
          venue: this.venue,
          filledSize: 0,
          avgPrice: 0,
          venueOrderId: `paper-${order.clientId}`,
          reason: "limit not marketable; resting",
        };
    }

    const fillPx = this.fillPrice(order, quote);
    let fillSize = order.size;
    if (order.reduceOnly && existing)
      fillSize = Math.min(order.size, existing.size);

    this.applyFill(order.symbol, order.side, fillSize, fillPx);

    return {
      order,
      status: "filled",
      venue: this.venue,
      filledSize: fillSize,
      avgPrice: fillPx,
      venueOrderId: `paper-${order.clientId}`,
      reason: "paper fill",
    };
  }

  private applyFill(
    symbol: string,
    side: Side,
    size: number,
    price: number,
  ): void {
    const existing = this.positions.get(symbol);
    const orderSigned = sideSign(side) * size;

    if (!existing || existing.size === 0) {
      this.positions.set(symbol, {
        symbol,
        venue: this.venue,
        side,
        size,
        entryPrice: price,
        markPrice: price,
        leverage: 1,
      });
      return;
    }

    const currentSigned = sideSign(existing.side) * existing.size;
    const newSigned = currentSigned + orderSigned;
    const sameDirection = currentSigned > 0 === orderSigned > 0;

    if (sameDirection) {
      const total = existing.size + size;
      existing.entryPrice =
        (existing.entryPrice * existing.size + price * size) / total;
      existing.size = total;
      existing.markPrice = price;
      return;
    }

    // opposite direction: realise PnL on the closed portion
    const closed = Math.min(existing.size, size);
    const pnl = sideSign(existing.side) * (price - existing.entryPrice) * closed;
    this.realizedPnl += pnl;
    this.cash += pnl;

    if (Math.abs(newSigned) < 1e-12) {
      this.positions.delete(symbol);
    } else if (newSigned > 0 === currentSigned > 0) {
      existing.size = Math.abs(newSigned);
      existing.markPrice = price;
    } else {
      existing.side = side;
      existing.size = Math.abs(newSigned);
      existing.entryPrice = price;
      existing.markPrice = price;
    }
  }

  closePosition(symbol: string): OrderResult {
    const pos = this.positions.get(symbol);
    if (!pos || pos.size === 0) {
      const dummy: Order = {
        symbol,
        side: "long",
        size: 1,
        orderType: "market",
        reduceOnly: true,
        tif: "gtc",
        clientId: "close",
      };
      return this.rejected(dummy, "no open position");
    }
    const order: Order = {
      symbol,
      side: opposite(pos.side),
      size: pos.size,
      orderType: "market",
      reduceOnly: true,
      tif: "gtc",
      clientId: `close-${symbol}`,
      note: "paper close",
    };
    return this.placeOrder(order);
  }

  private markPositions(): void {
    for (const pos of this.positions.values()) {
      const q = this.getQuote(pos.symbol);
      if (q) pos.markPrice = (q.bid + q.ask) / 2;
    }
  }

  getPositions(): Position[] {
    this.markPositions();
    return [...this.positions.values()].filter((p) => p.size > 0);
  }

  getAccount(): AccountState {
    const positions = this.getPositions();
    const unrealized = positions.reduce((s, p) => s + posUnrealizedPnl(p), 0);
    const used = positions.reduce((s, p) => s + posNotional(p), 0);
    const equity = this.cash + unrealized;
    return {
      venue: this.venue,
      equity,
      freeCollateral: Math.max(0, equity - used),
      positions,
    };
  }

  private rejected(order: Order, reason: string): OrderResult {
    return {
      order,
      status: "rejected",
      venue: this.venue,
      filledSize: 0,
      avgPrice: 0,
      reason,
    };
  }
}

File: sapphire-perps/src/sapphire_perps/venues/paper.py
"""In-memory paper broker.

A deterministic perp simulation used as the backing for any venue in paper
mode. It nets positions, applies configurable slippage, realises PnL on
reductions/flips, and reports a margin-style account snapshot.

It is intentionally simple (no funding, no liquidation engine) — enough to
exercise the full signal -> risk -> execution loop without touching real money.
"""

from __future__ import annotations

from ..types import (
    AccountState,
    Order,
    OrderResult,
    OrderStatus,
    OrderType,
    Position,
    Quote,
    Side,
)
from .base import PriceSource


class PaperBroker:
    def __init__(
        self,
        venue: str,
        price_source: PriceSource,
        starting_equity: float = 100_000.0,
        slippage_bps: float = 2.0,
    ):
        self.venue = venue
        self.price_source = price_source
        self.cash = float(starting_equity)
        self.realized_pnl = 0.0
        self.slippage_bps = slippage_bps
        self._positions: dict[str, Position] = {}

    # -- market data ---------------------------------------------------
    def get_quote(self, symbol: str) -> Quote | None:
        return self.price_source.quote(symbol)

    def _fill_price(self, order: Order, quote: Quote) -> float:
        if order.order_type is OrderType.LIMIT and order.limit_price is not None:
            base = order.limit_price
        else:
            base = quote.ask if order.side is Side.LONG else quote.bid
        slip = base * (self.slippage_bps / 1e4)
        # Slippage always works against the taker.
        return base + slip if order.side is Side.LONG else base - slip

    # -- execution -----------------------------------------------------
    def place_order(self, order: Order) -> OrderResult:
        quote = self.get_quote(order.symbol)
        if quote is None:
            return OrderResult(
                order=order,
                status=OrderStatus.REJECTED,
                venue=self.venue,
                reason=f"no price for {order.symbol}",
            )

        existing = self._positions.get(order.symbol)
        if order.reduce_only and existing is None:
            return OrderResult(
                order=order,
                status=OrderStatus.REJECTED,
                venue=self.venue,
                reason="reduce_only order with no open position",
            )

        # A limit order only executes if it's marketable. A buy must cross the
        # ask, a sell must cross the bid; otherwise it rests (PENDING) rather
        # than filling at an impossible price. We don't model a resting book, so
        # a non-marketable limit simply stays pending and applies no fill.
        if order.order_type is OrderType.LIMIT and order.limit_price is not None:
            marketable = (
                order.limit_price >= quote.ask
                if order.side is Side.LONG
                else order.limit_price <= quote.bid
            )
            if not marketable:
                return OrderResult(
                    order=order,
                    status=OrderStatus.PENDING,
                    venue=self.venue,
                    venue_order_id=f"paper-{order.client_id}",
                    reason="limit not marketable; resting",
                )

        fill_px = self._fill_price(order, quote)
        fill_size = order.size
        if order.reduce_only and existing is not None:
            fill_size = min(order.size, existing.size)

        self._apply_fill(order.symbol, order.side, fill_size, fill_px)

        return OrderResult(
            order=order,
            status=OrderStatus.FILLED,
            venue=self.venue,
            filled_size=fill_size,
            avg_price=fill_px,
            venue_order_id=f"paper-{order.client_id}",
            reason="paper fill",
        )

    def _apply_fill(self, symbol: str, side: Side, size: float, price: float) -> None:
        existing = self._positions.get(symbol)
        order_signed = side.sign * size

        if existing is None or existing.size == 0:
            self._positions[symbol] = Position(
                symbol=symbol,
                venue=self.venue,
                side=side,
                size=size,
                entry_price=price,
                mark_price=price,
            )
            return

        current_signed = existing.signed_size
        new_signed = current_signed + order_signed

        same_direction = (current_signed > 0) == (order_signed > 0)
        if same_direction:
            # Average up/down — no realised PnL.
            total = existing.size + size
            existing.entry_price = (
                existing.entry_price * existing.size + price * size
            ) / total
            existing.size = total
            existing.mark_price = price
            return

        # Opposite direction: realise PnL on the closed portion.
        closed = min(existing.size, size)
        self.realized_pnl += existing.side.sign * (price - existing.entry_price) * closed
        self.cash += existing.side.sign * (price - existing.entry_price) * closed

        if abs(new_signed) < 1e-12:
            del self._positions[symbol]
        elif (new_signed > 0) == (current_signed > 0):
            # Reduced but same side.
            existing.size = abs(new_signed)
            existing.mark_price = price
        else:
            # Flipped: leftover opens a fresh position at the fill price.
            existing.side = side
            existing.size = abs(new_signed)
            existing.entry_price = price
            existing.mark_price = price

    def close_position(self, symbol: str) -> OrderResult:
        pos = self._positions.get(symbol)
        if pos is None or pos.size == 0:
            # Build a zero-size dummy order to satisfy the result contract.
            dummy = Order(symbol=symbol, side=Side.LONG, size=1.0, reduce_only=True)
            return OrderResult(
                order=dummy,
                status=OrderStatus.REJECTED,
                venue=self.venue,
                reason="no open position",
            )
        order = Order(
            symbol=symbol,
            side=pos.side.opposite,
            size=pos.size,
            reduce_only=True,
            note="paper close",
        )
        return self.place_order(order)

    # -- account -------------------------------------------------------
    def _mark_positions(self) -> None:
        for pos in self._positions.values():
            q = self.get_quote(pos.symbol)
            if q is not None:
                pos.mark_price = q.mid

    def get_positions(self) -> list[Position]:
        self._mark_positions()
        return [p for p in self._positions.values() if p.size > 0]

    def get_account(self) -> AccountState:
        self._mark_positions()
        positions = self.get_positions()
        unrealized = sum(p.unrealized_pnl for p in positions)
        used_margin = sum(p.notional for p in positions)
        equity = self.cash + unrealized
        free = max(0.0, equity - used_margin)
        return AccountState(
            venue=self.venue,
            equity=equity,
            free_collateral=free,
            positions=positions,
        )
